components/TextFields/src/ContainedInputView/private/MDCTextControlStyleOutlined.m

```diff
@@ -0,0 +1,204 @@
+  if (![textControl isKindOfClass:[UIView class]]) {
+    [self removeStyleFrom:textControl];
+    return;
+  }
+  CGRect labelFrame = textControl.label.frame;
+  BOOL isLabelFloating = textControl.labelState == MDCTextControlLabelStateFloating;
+  CGFloat containerHeight = CGRectGetMaxY(textControl.containerFrame);
+  CGFloat lineWidth = (CGFloat)self.outlineLineWidths[@(textControl.textControlState)].doubleValue;
+  UIView *uiView = (UIView *)textControl;
+  [self applyStyleTo:uiView
+            labelFrame:labelFrame
+       containerHeight:containerHeight
+       isLabelFloating:isLabelFloating
+      outlineLineWidth:lineWidth];
+  self.outlinedSublayer.strokeColor =
+      ((UIColor *)self.outlineColors[@(textControl.textControlState)]).CGColor;
+}
+
+- (UIFont *)floatingFontWithNormalFont:(UIFont *)font {
+  CGFloat scaleFactor = kFilledFloatingLabelScaleFactor;
+  CGFloat floatingFontSize = font.pointSize * scaleFactor;
+  return [font fontWithSize:floatingFontSize];
+}
+
+- (void)removeStyleFrom:(id<MDCTextControl>)TextControl {
+  [self.outlinedSublayer removeFromSuperlayer];
+}
+
+- (id<MDCTextControlVerticalPositioningReference>)
+    positioningReferenceWithFloatingFontLineHeight:(CGFloat)floatingLabelHeight
+                              normalFontLineHeight:(CGFloat)normalFontLineHeight
+                                     textRowHeight:(CGFloat)textRowHeight
+                                  numberOfTextRows:(CGFloat)numberOfTextRows
+                                           density:(CGFloat)density
+                          preferredContainerHeight:(CGFloat)preferredContainerHeight {
+  return [[MDCTextControlVerticalPositioningReferenceOutlined alloc]
+      initWithFloatingFontLineHeight:floatingLabelHeight
+                normalFontLineHeight:normalFontLineHeight
+                       textRowHeight:textRowHeight
+                    numberOfTextRows:numberOfTextRows
+                             density:density
+            preferredContainerHeight:preferredContainerHeight];
+}
+
+#pragma mark Internal Styling Methods
+
+- (void)applyStyleTo:(UIView *)view
+          labelFrame:(CGRect)labelFrame
+     containerHeight:(CGFloat)containerHeight
+     isLabelFloating:(BOOL)isLabelFloating
+    outlineLineWidth:(CGFloat)outlineLineWidth {
+  UIBezierPath *path = [self outlinePathWithViewBounds:view.bounds
+                                            labelFrame:labelFrame
+                                       containerHeight:containerHeight
+                                             lineWidth:outlineLineWidth
+                                       isLabelFloating:isLabelFloating];
+  self.outlinedSublayer.path = path.CGPath;
+  self.outlinedSublayer.lineWidth = outlineLineWidth;
+  if (self.outlinedSublayer.superlayer != view.layer) {
+    [view.layer insertSublayer:self.outlinedSublayer atIndex:0];
+  }
+}
+
+- (UIBezierPath *)outlinePathWithViewBounds:(CGRect)viewBounds
```

Comment:
I like how these are pure functions without any state. We should consider enforcing that with the compiler somehow.

Suggestion: The only way I came up with was making this a UIBezierPath category also?
like this convenience initializer:
https://developer.apple.com/documentation/uikit/uibezierpath/1624368-bezierpathwithroundedrect?language=objc

Reply:
The method can also be made static `+ (UIBezierPath *)outlinePathWithViewBounds:(CGRect)viewBounds`

Reply:
I just made it a class method. I could also make it a category method, although I don't think it's strictly necessary..